File: 78-subsets/78-subsets.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        final = []
        res = []
        n = len(nums)
        def subset(start):
            final.append(res[:])
            for j in range(start,n):
                res.append(nums[j])
                subset(j+1)
                res.pop()
        subset(0)
        return final
        

File: 78-subsets/78-subsets.java
class Solution {
    private List<List<Integer>> res = new LinkedList<>();
    private LinkedList<Integer> temp = new LinkedList<>();
    private int start = 0;
    public List<List<Integer>> subsets(int[] nums) {
        res.add((List) temp.clone());
        for(int i = start; i < nums.length; i++){
            start = i + 1;
            temp.addLast(nums[i]);
            subsets(nums);
            temp.pollLast();
        }
        return res;
    }
}